### PR TITLE
Various Documentation improvements

### PR DESCRIPTION
**Pull request by jaxoncreed**
Covers:
 - Describes the use case and implementation of PoP tokens
 - The description of asking for consent on the example page is more in line with reality
 - Extremely detailed walkthrough

### CODE REVIEW

File: README.md
Comment (michielbdejong) [inline README.md]: ".. to let the client access a wide .."

File: README.md
Comment (jaxoncreed) [inline README.md]: Mention that these keys are short lived

File: README.md
Comment (michielbdejong) [inline README.md]: yeilds -> yields

**Review by michielbdejong — COMMENTED** [review]
in the png, '6. Generates and id_token' -> '6. Generates an id_token'
step 5 of 'sending request' seems incorrect to me, IIUC RP only contacts OP to retrieve their jwks, not to ask them to perform validation.

**Comment by zenomt** [comment]
is this PR just to document how the reference implementation currently works, or to also suggest new functionality that is yet to be implemented?  i see there's mention (in "README.md -> The Solution, step 3"), of the origin of the redirect_uri becoming "the `aud`ience".

i haven't checked the source to see if that's what the current NSS OP actually does, but regardless that has some issues (most importantly, the `aud` MUST include the client_id, which as discussed in #12 the client_id can't be just the redirect_uri origin, or even the redirect_uri).  i think at least that part should be called out as "as currently implemented" or "proposed behavior" so that it can be commented on appropriately.

if this is documenting what's currently implemented, or at least what's currently designed/planned, then discussing issues with it isn't appropriate in this PR's conversation, and should be in an actual Issue once this change is merged. however, if this PR is also proposing new behavior, then discussing issues with the new behavior would be appropriate here.

**Comment by michielbdejong** [comment]
@jaxoncreed ping

**Comment by jaxoncreed** [comment]
I've now finished the detailed application workflow, so I think we're good to go. It's not the 100% rewrite this needs, but it provides some much better context.

**Review by michielbdejong — DISMISSED** [review]
I didn't read/verify all of it but looks like a great improvement over the current text. We can always iterate if we discover errata. A second review from @zenomt and/or @dmitrizagidulin would be helpful!

**Comment by michielbdejong** [comment]
We should probably update the list of people with write access to this repo to match the list of people who know most about how webid-oidc works. Created https://github.com/solid/webid-oidc-spec/issues/32 in relation to this.

**Comment by michielbdejong** [comment]
@kjetilk @Mitzi-Laszlo @timbl @justinwb @RubenVerborgh: two of you need to approve this before it can be merged.

**Comment by RubenVerborgh** [comment]
@michielbdejong Will do; useful to use the "assign reviewers" functionality for that.

File: README.md
Comment (RubenVerborgh) [inline README.md]: Perhaps say a word about what they are, or at least expand acronym?

File: README.md
Comment (RubenVerborgh) [inline README.md]: But we are using a standard, right? Important that people understand this is not a custom extension.

File: application-workflow-detailed.md
Comment (RubenVerborgh) [inline application-workflow-detailed.md]: invalid JSON

**Review by RubenVerborgh — DISMISSED** [review]
Technically correct, but I feel that we are too much describing an existing spec here. We should point more and explain less (unless the actual specs are not clear). However, good from a documentation perspective, but need to revisited when turning into an actual spec, where we want to avoid repetition of other specs.

**Comment by zenomt** [comment]
in the current version of the sequence diagram, authorization steps 3 and 4 (request OP configuration) are to Alice's POD, but i think you mean for them to go to her OP.

File: application-workflow-detailed.md
Comment (justinwb) [inline application-workflow-detailed.md]: - typo - should be "trying to access photos stored in Bob's Pod"
- need to delineate between app and webapp. the link is to add trusted origins, but it says that bob indicated through access control that alice could access his photo using any "app".

File: application-workflow-detailed.md
Comment (justinwb) [inline application-workflow-detailed.md]: typo - thrid -> third

File: application-workflow-detailed.md
Comment (justinwb) [inline application-workflow-detailed.md]: typo: naviages -> navigates

File: application-workflow-detailed.md
Comment (justinwb) [inline application-workflow-detailed.md]: typo: via he -> via her
typo: returns and -> returns an

File: application-workflow-detailed.md
Comment (justinwb) [inline application-workflow-detailed.md]: typo: either provider -> either provide
typo: or the service the url -> or the url

File: application-workflow-detailed.md
Comment (justinwb) [inline application-workflow-detailed.md]: typo: can have acces to -> can have access to
typo: Solid has it's -> Solid has its

File: application-workflow-detailed.md
Comment (justinwb) [inline application-workflow-detailed.md]: typo: it's login screen -> its login screen

File: application-workflow-detailed.md
Comment (justinwb) [inline application-workflow-detailed.md]: reword this -> "ways to communicate what the kind of information the authorization request should wants to get access to"

**Comment by kjetilk** [comment]
So, I gotta admit that I do not feel quite up to the task of reviewing this for technical accuracy. I think that once the rest of you have worked out the details, I can review it to see if I now understand the documentation, and then, I can merge if I do. OK?

**Comment by zenomt** [comment]
also in the ApplicationWorkflow.png diagram, in the "SENDING REQUEST" section, the steps where Bob's Pod loads Alice's WebID document (to check that Alice's OP is an authorized issuer) and loads Alice's OP's openid-configuration file (so that "5. Requests public keys" can happen), are missing.

**Comment by jaxoncreed** [comment]
@zenomt The request to Alice's Pod was on purpose. Both the Pod and the IDP should provide a well-known/openid-configuration so that clients referencing the Pod can know how to log in even if the IDP is on a separate server.

Though it might be better to do this through a `Link` header. What do you think?

**Comment by jaxoncreed** [comment]
> also in the ApplicationWorkflow.png diagram, in the "SENDING REQUEST" section, the steps where Bob's Pod loads Alice's WebID document (to check that Alice's OP is an authorized issuer) and loads Alice's OP's openid-configuration file (so that "5. Requests public keys" can happen), are missing.

Good point that should be added

**Comment by zenomt** [comment]
> The request to Alice's Pod was on purpose. Both the Pod and the IDP should provide a well-known/openid-configuration so that clients referencing the Pod can know how to log in even if the IDP is on a separate server.

sorry, i don't understand. if the Pod isn't an OpenID Provider, why does it need an openid-configuration? are you sure you didn't mean "if Alice enters her WebID, instead of her OP in step 2, load Alice's WebID Profile to discover her `solid:oidcIssuer`"?  the user interface for logging in directly in your browser is HTML in the `401` response body, and you use a bearer token to access a restricted resource programmatically.

also, i think the main reason for my confusion is that step 2 says "Alice selects her OP", and then step 3 is "request OP configuration from Alice's Pod". that should only happen if Alice selected her Pod as her OP; otherwise RP wouldn't know what Alice's Pod is yet.

regardless, if steps 3 and 4 aren't to Alice's OP, then there also needs to be a transaction before 7 & 8 to load the openid-configuration from Alice's OP.

> Though it might be better to do this through a `Link` header. What do you think?

i'm not a fan of a `Link` header to discover the authorized issuer; please see #18 for a full discussion on that.

File: application-workflow-detailed.md
Comment (jaxoncreed) [reply application-workflow-detailed.md]: For clarification, I changed all references to "app" to "webapp"

**Comment by zenomt** [comment]
> sorry, i don't understand.

the other situation that comes to mind that you might mean is in the case today where solid.community has a wildcard DNS entry to cover all identities *.solid.community, and answers the same openid-configuration file no matter what `Host`/subdomain was requested.

when that's the case, if Alice enters the origin for her webid (which is probably also her pod) instead of her OP URI, RP can ask for `https://alice.solid.community/.well-known/openid-configuration`, and will get back a configuration that's actually for `https://solid.community`.  in particular, the `issuer` claim in the configuration will say `https://solid.community`, not `https://alice.solid.community`.  when that mismatch happens, you need to re-fetch the configuration from the real issuer, to make sure that Alice isn't lying.  otherwise it might be possible for issuer `https://alice.solid.community` to list different endpoints (especially `authorization_endpoint`, `token_endpoint`, and `jwks_uri`) and to issue a counterfeit id_token whose `iss` is `https://solid.community`, which, depending on who received it when, might be believed.

**Comment by jaxoncreed** [comment]
@zenomt You were right the first time. I checked it out and found 2 ways to pick out the issuer from the profile. So I will document them here and remove the requirement for the RS to have a /.well-known/openid-configuration

**Comment by zenomt** [comment]
@jaxoncreed the current revision of ApplicationWorkflow.png AUTHORIZATION step 3 still goes to Alice's Pod, even though the response comes from Alice's OP. i think you just missed that one arc. also 14 "and" should be "an".

to keep the same level of detail as in the rest of the diagram, in "SENDING REQUEST", between box 6 and fetch 7 should be another fetch for the OP configuration by Bob's Pod, so it knows how to request the public keys.  one could argue that "7 requests public keys" means "do the necessary steps to request the public keys", but since it involves 2 round trips, i think it should be called out separately.

**Comment by jaxoncreed** [comment]
@zenomt  right you are

File: application-workflow-detailed.md
Comment (TallTed) [inline application-workflow-detailed.md]: typos...
`thrid` -> `third`
`authroization` -> `authorization`

File: application-workflow-detailed.md
Comment (michielbdejong) [reply application-workflow-detailed.md]: can you explain why it is invalid?  I just tried this:
```js
console.log(JSON.parse(`{\
  "grant_types": ["implicit"],\
  "issuer": "https://secureauth.example",\
  "redirect_uris": ["https://www.decentphotos.example/"],\
  "response_types": ["id_token token"],\
  "scope": "openid profile"\
}`))
```
output:
```js
{ grant_types: [ 'implicit' ],
  issuer: 'https://secureauth.example',
  redirect_uris: [ 'https://www.decentphotos.example/' ],
  response_types: [ 'id_token token' ],
  scope: 'openid profile' }
```

File: application-workflow-detailed.md
Comment (TallTed) [inline application-workflow-detailed.md]: `[oidc scpes]`
-> `[OIDC scopes]`

File: application-workflow-detailed.md
Comment (TallTed) [inline application-workflow-detailed.md]: `webId-oidc`
-> `WebID-OIDC`

File: application-workflow-detailed.md
Comment (TallTed) [inline application-workflow-detailed.md]: hopefully clearer...
`WebId`
-> `WebID`

**Comment by jaxoncreed** [comment]
Can we resurrect this and approve it @kjetilk. I keep sending links to this document because people keep asking how the login works and can't find this.

**Comment by kjetilk** [comment]
> Can we resurrect this and approve it @kjetilk. I keep sending links to this document because people keep asking how the login works and can't find this.

Right! But as of now, it has no approving reviews... As I said above, I'd prefer that others provide reviews and then I review and merge it, because I don't feel up to the task of performing a technical review myself. So, @justinwb , @RubenVerborgh , @TallTed , @michielbdejong , please have another look and update your reviews.

**Comment by RubenVerborgh** [comment]
I can't; not knowledgeable enough about OIDC.

**Comment by kjetilk** [comment]
My problem too... I think the [Authentication Panel](https://github.com/solid/culture/blob/master/panels.md#authentication-panel) should convene urgently to resolve this.

**Comment by dmitrizagidulin** [comment]
@kjetilk 
> I think the Authentication Panel should convene urgently to resolve this.

I agree!

**Review by dmitrizagidulin — DISMISSED** [review]
Read through this carefully on a call with @jaxoncreed.

Looks great, thank you SO much for doing this work, Jackson!

File: application-workflow-detailed.md
Comment (TallTed) [inline application-workflow-detailed.md]:
```
`scope=open_id`
``` 
became
```
`scope=openid`
```
so
```
`open_id` is a scope
```
should become
```
`openid` is a scope
```

**Comment by justinwb** [comment]
Should link to the new document at application-workflow-detailed.md, but the links to this and application-workflow.md should be put in context based on the use case. File names may also need to be changed to align with that and prevent confusion.

**Review by justinwb — APPROVED** [review]
Nice improvement to the current state of the documentation. There's still content required to provide full coverage, especially as it relates to the different patterns of use and how to apply them, but I think that this moves the ball forward in this current instantiation of the spec. The rest of the efforts should be channeled towards https://github.com/solid/specification. With that in mind, I'm good with approving these changes here.

**Review by dmitrizagidulin — APPROVED** [review]

**Pull request merged**
